### ISSUE TITLE
putty-cac/pageant does not support FIPS compliant based keys

### ISSUE THREAD

**Issue by karlg100** [NoMoreFood/putty-cac#63]
With FIPS mode enabled on RHEL8, pageant PKI generated keys are rejected.  ssh-rsa keys are now considered not FIPS compliant.

to reproduce:
* ssh in using pki generated keys, verify working
* enable a RHEL8 system to FIPS mode : fips-mode-setup –enable
* reboot the RHEL8 system
* attempt to ssh in again with the same key

the following error is generated in the ssh debug logs:
sshd[10632]: userauth_pubkey: key type ssh-rsa not in PubkeyAcceptedKeyTypes [preauth]

A workaround is to enable ssh-rsa in the list of acceptedKeyTypes (doing so beaks FIPS compliance), but with CAPI, another error is generated relating to the signing algorithm also out of compliance. Using ActivID's PKCS11 however bypasses this second issue.

Redhat has documented this re Putty and FIPS, but the solution is to regenerate keys which is not possible with PKI certs.
https://access.redhat.com/solutions/4906221

**Comment by NoMoreFood** [comment]
I will investigate this more this coming weekend.

**Comment by chuckmilam** [comment]
@karlg100 I ran into a similar issue, not sure if it's the same, but figure I'd at least mention it for you to try.  

I was getting this error from PuTTY/PuTTY-CAC when I'd attempt to connect:

> "Couldn't agree a host key algorithm (available: rsa-sha2-512, rsa-sha2-256)."

When I updated my RHEL 8 systems per the DISA STIG, specifically the checks requiring the system "....only use ciphers employing FIPS 140-2-approved algorithms...." where modification of files in /etc/crypto-policies/back-ends/ is required, things starting working with PKI (smartcard) certs again.  Maybe give that a look?

**Comment by karlg100** [comment]
Thanks for the tip.

So did see that at one point... (can’t remember where in my testing) and will go back and check for host key issues.  That may have been with CAPI libraries.

But I know that with DEBUG3 level enabled on sshd, you can clearly see putty-cac is passing an ssh-rsa type authorized key line, which is for sure not allowed & disabled for FIPS support.

Karl

**Comment by NoMoreFood** [comment]
I am waiting for the 0.75 release of upstream PuTTY which addresses this issue (see https://www.chiark.greenend.org.uk/~sgtatham/putty/).  I anticipate a release a PuTTY CAC within a week or two since this release changes two major parts of how PuTTY CAC integrates into PuTTY.

**Comment by NoMoreFood** [comment]
The new upstream version was released.  Please let me know if you'd be willing to test once my changes are done.

**Comment by NoMoreFood** [comment]
@karlg100 @chuckmilam Can you see if the binaries provided at https://github.com/NoMoreFood/putty-cac/tree/master/binaries resolve the issue? I haven't setup a FIPS compliant Linux box, but I will if it looks like you're still seeing issues.

**Comment by karlg100** [comment]
I'll attempt to test Monday.  I have a call with my customer and we will see if we can test against a STIGed RHEL8 box.

**Comment by chuckmilam** [comment]
Began testing this morning.  Installed with 64-bit MSI installer.  SSH connections to FIPS-enabled RHEL 8 and RHEL 7 boxes seem OK.  

Running Ansible plays on a test system for full STIG compliance now.

Observed changes:

1. Pagent will now exit seemingly randomly and have to be restarted, which is not expected behavior and is a change from v0.74.  I used to start pagent at the start of the day and use it all day long.  I will have to test more to see if there's a pattern here.
2. ssh-agent on RHEL 7 host used for Ansible is now losing the smartcard CAPI certs and struggling to add other SSH identities at the local command line. I'll have to experiment more with this, including using RHEL 8 systems, and attempt to see if there's a pattern here.

**Comment by karlg100** [comment]
We tested v0.75 (64bit) with both CAPI and PKCS11 with ActivID, and it's still presenting ssh-rsa based host keys. Did we miss a config option here?

**Comment by chuckmilam** [comment]
@karlg100 using CAPI here.  Only doing the Red Hat STIG kickstart addon "org_fedora_oscap" with some follow-up Ansible to finish the required settings per the RHEL 8 STIG.  Test system in is FIPS mode, as defined at install time and confirmed with multiple checks:  fipscheck, grub boot arguments, and /proc/sys/crypto/fips_enabled.

This is a stand-alone system, not AD joined or anything of the sort (yet.)  Local user authorized_keys are captured from CAC/PIV smartcards with PuTTY-CAC and copied into place with Ansible, configured per the _DoD PKE Reference Guide._  

Haven't seen the problems you're seeing here,  We must have some different config setting somewhere in the mix.

**Comment by NoMoreFood** [comment]
@chuckmilam If there’s any case you can make out to reproduce the issue(s), it would be appreciated.  Pageant doesn’t do much when its idling so my guess is some operation is triggering a crash. Thank you for testing.

**Comment by karlg100** [comment]
@chuckmilam - this should be checked before the ssh_authroized_keys is looked up/derived from the LDAP provider.

Check /etc/crypto-policies/back-ends/openssh.config.  this should be overriding the sshd_config with the acceptable KeyAlgorithems and PubkeyAcceptedKeyTypes.  check that ssh-rsa is not there.  AFAIK, this is default for RHEL8 FIPS mode to lock down the crypto when FIPS is enabled.

Also, crank up sshd's LogLevel to DEBUG3 and watch the journalctl -af output.  Look for the authorized_keys string that's sent from Pageant, and see if it's ssh-rsa.

**Comment by chuckmilam** [comment]
@karlg100 I've verified ssh-rsa is not in /etc/crypto-policies/back-ends/openssh.config. I do see ssh-rsa in the journalctl -af output when authenticating with a smartcard cert.  I'm going to go back and verify all STIG-required FIPS and crypto settings on this test system.

**Comment by chuckmilam** [comment]
> Pageant doesn’t do much when its idling so my guess is some operation is triggering a crash. 

@NoMoreFood No pagent issues at all today on my test Win 10 system.  Probably something else going on with all the various other mandatory compliance agents on this system.

**Comment by karlg100** [comment]
Check/verify that somewhere in the Sshd_config that ssh-rsa isn’t creeping in.  rsa-ssh should not be accepted on a FIPS RHEL8 system.

Karl Grindley

**Comment by NoMoreFood** [comment]
@karlg100 Are you actually seeing failures? I believe you should still expect to see rsa-ssh in some cases.  See https://levelup.gitconnected.com/demystifying-ssh-rsa-in-openssh-deprecation-notice-22feb1b52acd.

**Comment by karlg100** [comment]
yes.  Note this is not an SSH change, this is a Redhat implementation for FIPS compliance.

```
May 11 05:41:13 host sshd[1071599]: debug2: input_userauth_request: try method publickey [preauth]
May 11 05:41:13 host sshd[1071599]: debug2: userauth_pubkey: valid user root querying public key ssh-rsa [KEY REMOVED]
May 11 05:41:13 host sshd[1071599]: userauth_pubkey: key type ssh-rsa not in PubkeyAcceptedKeyTypes [preauth]
```

This should not be hard to replicate.

* load a patched RHEL8.3 system
* configure for your smartcard settings (test sshing in)
* enable FIPS mode `fips-mode-setup --enable`
* enable LogLevel DEBUG2 or DEBUG3 in /etc/ssh/sshd_config
* reboot
* try to SSH in with a smartcard

https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/security_hardening/using-the-system-wide-cryptographic-policies_security-hardening

this breaks because the crypto polices for SSH are explicitly set in 
/etc/crypto-policies/back-ends/opensshserver.config. The PubkeyAcceptedKeyTypes does not include rsh-ssh. This file is referenced in the systemd unit file /usr/lib/systemd/system/sshd.service and used when sshd is started.

When FIPS mode is enabled, the opensshserver.config has the following flags set to pass directly to sshd when it is started.  This overrides all other configurations in sshd_config and compile time defaults.

```
# cat /etc/crypto-policies/back-ends/opensshserver.config
CRYPTO_POLICY='-oCiphers=aes256-gcm@openssh.com,aes256-ctr,aes256-cbc,aes128-gcm@openssh.com,aes128-ctr,aes128-cbc -oMACs=hmac-sha2-256-etm@openssh.com,hmac-sha1-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha2-256,hmac-sha1,hmac-sha2-512 -oGSSAPIKeyExchange=no -oKexAlgorithms=ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512 -oHostKeyAlgorithms=ecdsa-sha2-nistp256,ecdsa-sha2-nistp256-cert-v01@openssh.com,ecdsa-sha2-nistp384,ecdsa-sha2-nistp384-cert-v01@openssh.com,ecdsa-sha2-nistp521,ecdsa-sha2-nistp521-cert-v01@openssh.com,rsa-sha2-256,rsa-sha2-256-cert-v01@openssh.com,rsa-sha2-512,rsa-sha2-512-cert-v01@openssh.com -oPubkeyAcceptedKeyTypes=ecdsa-sha2-nistp256,ecdsa-sha2-nistp256-cert-v01@openssh.com,ecdsa-sha2-nistp384,ecdsa-sha2-nistp384-cert-v01@openssh.com,ecdsa-sha2-nistp521,ecdsa-sha2-nistp521-cert-v01@openssh.com,rsa-sha2-256,rsa-sha2-256-cert-v01@openssh.com,rsa-sha2-512,rsa-sha2-512-cert-v01@openssh.com -oCASignatureAlgorithms=ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,rsa-sha2-256,rsa-sha2-512'

```

**Comment by NoMoreFood** [comment]
@karlg100 Okay, can you verify you *don't* see the error when simply using a standard PuTTY key?

**Comment by karlg100** [comment]
Each distro and version have implemented FIPS slightly different.  Much of this is due to the FIPS requirements at the time for which each maintainer configured compliance. I'm not sure moving CRYPTO_POLICY over to Ubuntu will work, as I've not gone down the code path to how Ubuntu implements FIPS. (note that to get the full FIPS modules on Ubuntu, you must have a paid entitlement/support from Canonical). RHEL7 FIPS mode is not the same as RHEL8 for reasons above.

If you're having trouble reproducing on RHEL8, can you please post the process you're using to configure your system post media/iso install?


As for testing on our side, we tested the available options in 0.75. Only ECDSA key type worked.

No go with RSA: (as we've been seeing)
`May 13 12:46:30 vls8sssdtest sshd[2928]: userauth_pubkey: key type ssh-rsa not in PubkeyAcceptedKeyTypes [preauth]
`

No go with ED25519:
`May 13 12:23:00 vls8sssdtest sshd[3706]: userauth_pubkey: key type ssh-ed25519 not in PubkeyAcceptedKeyTypes [preauth]
`

but ECDSA works!
`May 13 12:18:51 vls8sssdtest sshd[3421]: Accepted publickey for [user] from [ip] port 61357 ssh2: ECDSA SHA256:[hash]
`

**Comment by NoMoreFood** [comment]
This is what I see:

```
May 14 14:42:32 localhost sshd[21220]: debug2: userauth_pubkey: valid user [user] querying public key rsa-sha2-512 [pubkey] [preauth]
May 14 14:42:32 localhost sshd[21220]: debug1: userauth_pubkey: test pkalg rsa-sha2-512 pkblob RSA SHA256:[hash] [preauth]
```
followed by:

```
May 14 14:42:32 localhost sshd[21220]: Accepted key RSA SHA256:[hash] found at /home/[user]/.ssh/authorized_keys:2
```

**Comment by chuckmilam** [comment]
> If you're having trouble reproducing on RHEL8, can you please post the process you're using to configure your system post media/iso install?

Configurations were set following this STIG guidance, using Ansible now:  

https://www.stigviewer.com/stig/red_hat_enterprise_linux_8/2020-11-25/finding/V-230251

**Comment by karlg100** [comment]
hmm, ok. so now I'm wondering if it's a pageant/windows issue.  are you running FIPS mode on the windows side?

**Comment by chuckmilam** [comment]
@karlg100  Flipped the FIPS mode registry key on my Windows 10 test system.  No longer see ssh-rsa in /var/log/secure. No problems connecting via pagent and smartcard. Working as expected here.

**Comment by NoMoreFood** [comment]
So where'd we land on this?  Do we still think there's an outstanding issues?

**Comment by Racheakt** [comment]
I happen to be working with a FIPS compliant RHEL 8 migration. 

I have downloaded the latest Putty CaC and it is still generating ssh-rsa keys from the CaC (both capi and pkcs)

Basic setup for RHEL 8 instance was to select the 8 STIG profile during the install, and when done run the DISA Ansible STIG playbook. Reboot.

https://access.redhat.com/solutions/4906221


> Environment
> 
>     Red Hat Enterprise Linux 8
>         FIPS
>     PuTTY
> 
> Issue
> 
>     Connecting using PuTTY to a RHEL 8 system running FIPS-140-2, the connection fails, as reported in /var/log/secure
>     Raw
> 
>     
> 
> > TIMESTAMP sshd[PID]: userauth_pubkey: key type ssh-rsa not in PubkeyAcceptedKeyTypes [preauth]
> 
>     Connection using PuTTY to a RHEL 7 system running FIPS-140-2 works fine
> 
> Resolution
> 
> This is expected behaviour: ssh-rsa keys are not FIPS-140-2 compliant.
> 
>     Generate a new key which is compliant with FIPS-140-2, for example ECDSA with curve nistp256.
>     Add the public key to authorized_keys file on the destination system
> 
> Root Cause
> 
>     Default PuTTY key algorithm is ssh-rsa, which isn't FIPS-140-2 compliant
>     With RHEL 7, this type of keys were tolerated, but not anymore with RHEL 8


Basically ssh-rsa generated keys out going forward. 

I was able to use puttygen to generate a ppk a compliant ecdsa-sha2-nistp521 keyfile, however using the CaC  with CAPI and PCKS options still generate ssh-rsa key for the authorized_key file. 

So at this time Putty-CaC is not working for me at least to connect to a FIPS-2 Complaint RHEL instance using the CaC.

**Comment by NoMoreFood** [comment]
@Racheakt Yeah, sounds maybe just like a binary issue.   Closing issue.  Thanks!

**Comment by Racheakt** [comment]
> 
> 
> @Racheakt Yeah, sounds maybe just like a binary issue. Closing issue. Thanks!

No problem, thanks for your work on this.
